PE-426: Change core app over to Kotlin

lucianocheng opened the pull request:
Ports the core PhotoEditorSDK over to Kotlin.

lucianocheng commented:
@burhanrashid52 not yet ready to review. Still actively working on it.

burhanrashid52 commented:
@lucianocheng I've fixed the conflicts in `kotlin-conversion` and merged it to master. You can now take the latest from master and can point this branch to master.

lucianocheng commented:
@burhanrashid52 perfect timing, we are almost done.

lucianocheng reviewed (commented):
@tiendung717 flagged some minor issues, please address. Thanks!

tiendung717 commented on app/src/main/java/com/burhanrashid52/photoediting/base/BaseFragment.kt:
```diff
@@ -11,14 +11,4 @@ import androidx.fragment.app.Fragment
  */
-abstract class BaseFragment : Fragment() {
-    protected abstract val layoutId: Int
-    override fun onCreateView(
```
Move layout id to Fragment constructor (Kotlin style). Don't need to call function again

tiendung717 commented on app/src/main/java/com/burhanrashid52/photoediting/EditImageActivity.kt:
```diff
@@ -100,11 +103,13 @@ class EditImageActivity : BaseActivity(), OnPhotoEditorListener, View.OnClickLis
-            //.setDefaultTextTypeface(mTextRobotoTf)
-            //.setDefaultEmojiTypeface(mEmojiTypeFace)
-            .build() // build photo editor sdk
+        mPhotoEditor = mPhotoEditorView?.run {
```
I used null safe type (?) to execute code then the scope is not null and we don't need to call view again

tiendung717 commented on app/src/main/java/com/burhanrashid52/photoediting/EmojiBSFragment.kt:
```diff
@@ -59,8 +59,10 @@ class EmojiBSFragment : BottomSheetDialogFragment() {
     }
 
     inner class EmojiAdapter : RecyclerView.Adapter<EmojiAdapter.ViewHolder>() {
+        var emojisList = getEmojis(activity)
```
Removed variable and used in companion

tiendung717 commented on app/src/main/java/com/burhanrashid52/photoediting/StickerBSFragment.kt:
```diff
@@ -52,6 +52,8 @@ class StickerBSFragment : BottomSheetDialogFragment() {
         (contentView.parent as View).setBackgroundColor(resources.getColor(android.R.color.transparent))
         val rvEmoji: RecyclerView = contentView.findViewById(R.id.rvEmoji)
+
+        // TODO(lucianocheng): Move gridLayoutManager to xml file.
```
Removed

tiendung717 commented on photoeditor/src/main/java/ja/burhanrashid52/photoeditor/CustomEffect.kt:
```diff
@@ -0,0 +1,70 @@
+        }
+    }
+
+    // TODO(cheng): Implement Builder Pattern
```
Removed

tiendung717 commented on photoeditor/src/main/java/ja/burhanrashid52/photoeditor/MultiTouchListener.kt:
```diff
@@ -0,0 +1,267 @@
+                view.bringToFront()
+                firePhotoEditorSDKListener(view, true)
+            }
+            MotionEvent.ACTION_MOVE ->                 // Only enable dragging on focused stickers.
```
Updated

tiendung717 commented on photoeditor/src/main/java/ja/burhanrashid52/photoeditor/PhotoFilter.kt:
```diff
@@ -0,0 +1,19 @@
+ * @since 2/14/2018
+ */
+enum class PhotoFilter {
+    NONE, AUTO_FIX, BLACK_WHITE, BRIGHTNESS, CONTRAST, CROSS_PROCESS, DOCUMENTARY, DUE_TONE, FILL_LIGHT, FISH_EYE, FLIP_VERTICAL, FLIP_HORIZONTAL, GRAIN, GRAY_SCALE, LOMISH, NEGATIVE, POSTERIZE, ROTATE, SATURATE, SEPIA, SHARPEN, TEMPERATURE, TINT, VIGNETTE
```
Updated

tiendung717 commented on photoeditor/src/main/java/ja/burhanrashid52/photoeditor/TextStyleBuilder.kt:
```diff
@@ -0,0 +1,265 @@
+     * Enum to maintain current supported style properties used on on [PhotoEditor.addText] and [PhotoEditor.editText]
+     */
+    enum class TextStyle(val property: String) {
+        SIZE("TextSize"), COLOR("TextColor"), GRAVITY("Gravity"), FONT_FAMILY("FontFamily"), BACKGROUND(
```
Updated

lucianocheng commented:
@burhanrashid52 I think this is ready.

Let me know if you see anything that needs changing, otherwise please merge in. Thanks.

burhanrashid52 commented on photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt:
```diff
@@ -0,0 +1,217 @@
+    private fun createShape() {
+        val shape: AbstractShape
+        var paint = createPaint()
+        when {
```
Can we use the when on the shapeType so that we don't need to explicitly in condition 
```
when(currentShapeBuilder?.shapeType){
   ShapeType.OVAL -> {}
}
```

lucianocheng replied:
@burhanrashid52 so I may be reading it wrong, but would that mean we would need to move the `isErasing` condition out of the `when` block?

burhanrashid52 replied:
Yes @lucianocheng

lucianocheng replied:
So something like this to manage the `else` `BrushShape` condition? @burhanrashid52 

```
shape = BrushShape()
if (isErasing) {
  paint = createEraserPaint()
} else {
  when(currentShapeBuilder?.shapeType){
     ShapeType.OVAL -> {
         shape = OvalShape()
      }
  }
  ...
}
```

burhanrashid52 replied:
Yes @lucianocheng

lucianocheng replied:
👍 addressed.

burhanrashid52 reviewed (approved):
First of all, Thank you for taking the time to convert this into Kotlin. Overall I am happy with most of the changes, just added one comment.
Since it's now in Kotlin I see many opportunities to do some cleanup and make the API design better. Maybe a future task to work on.

lucianocheng commented:
@burhanrashid52 comments addressed. I think we're set 👍 .

> First of all, Thank you for taking the time to convert this into Kotlin. Overall I am happy with most of the changes, just added one comment.

👍 . Shout out to @tiendung717 on my team who did the bulk of the core library conversion.

> Since it's now in Kotlin I see many opportunities to do some cleanup and make the API design better. Maybe a future task to work on.

👍 . Now that the code is converted to Kotlin, here's our punch list for things we will port over. We already have solutions for these:

1. Fixes for file saving / permission logic
2. Fixed sticker handle sizes
3. Better handling of pinch action for resizing.

Re: API / Further work. Speaking strictly for our use case, the most important things to add are also the hardest (and have been on the PhotoEditor TODO list for awhile). We have sketches for these but they will require significant refactoring, we will try them ourselves first if we get to them:

1. Zoom / crop / rotation of the main image (requires refactoring the main image stack).
2. Ability to save a draft of the edit, and recover a draft (requires tracking and recovering all state).
3. Ability to "represent" a high resolution version of the image rather than saving from the Views (requires (2) and rewriting the export logic).

burhanrashid52 commented:
Thank you @tiendung717 for your effort.

@lucianocheng Sounds like a good plan to me. Let's me first release this version 2.0, meanwhile, you can raise PR for the already fix bugs.

After that, we can discuss more on the refactoring part.
